### DISCUSSION TITLE
Maplibre latest version combination + minification with terser break the code

### DISCUSSION THREAD

**Discussion by andreapernici** [maplibre/maplibre-gl-js#4923] [Q&A]
I tested various configuration of terser in order to prevent maplibre breaking, but unfortunately not figuring out.

Is there some suggestion of terser option to prevent maplibre latest version from breaking?

In dev mode where we don't minify the code everything works with maplibre-gl combined with other files, but in production mode after the minification everything stop working without any error in console so it's very difficult to debug.

If someone know of the terser configuration in order to prevent breaking maplibre.

Thanks

**Comment by HarelM** [comment]
It is not clear what the problem is, I'm not very familiar with terser, but I can't think of a major change in the code that could've break stuff that worked...
Please provide a more detailed problem definitions and steps to reproduce.

**Comment by andreapernici** [comment]
Now I combine

- maplibre-gl.js
- myfile-1.js
- myfile-2.js

in order to load everything async.

After combining everything is fine.

When instead doing the minification process on the combined file with terser and gulp the maplibre script stop working without providing any error on the browser side.

Have no more info on why, but I tested combining and minifing

- myfile1.js
- myfile2.js 

and then injecting on top the maplibre-gl.js file to exclude issue on myfile1 and 2 in order to isolate the issue.

So the issue is in the minification process of the maplibre-gl.js.

Can it be due to minifing an already minified file?

**Comment by HarelM** [comment]
What is the last version that worked? What commandline are you running? What if you don't add any code and just try the maplibre-gl.js file? What about using the -dev.js version as it is not minified?

**Comment by andreapernici** [comment]
The latest working versione is

MapLibre GL JS is licensed under the 3-Clause BSD License. Full text of license: https://github.com/maplibre/maplibre-gl-js/blob/v1.15.2/LICENSE.txt

Already tested -dev and -unminified

The Gulp task is do the following

```
gulp
        .src(files)
        .pipe(concat(name + ".js"))
        .pipe(plumber())
        .pipe(
          babel({
            presets: [
              [
                "@babel/preset-env",
                {
                  modules: false,
                  exclude: [
                    "@babel/plugin-transform-regenerator",
                    "@babel/plugin-transform-async-to-generator",
                    "@babel/plugin-transform-typeof-symbol",
                  ],
                },
              ],
            ],
          })
        )
        .pipe(
          terser({
            format: {
              comments: false,
            },
          })
        )
        .on("error", function (err) {
          console.log("errore");
          console.log(err);
        })
        .pipe(rev())
        .pipe(rename({ extname: ".min.js" }))
        .pipe(gzip({ append: true, gzipOptions: { level: 9 } }))
        .pipe(gulp.dest(config.static_js_dest));

```

**Comment by HarelM** [comment]
This is a version from 3 year ago.
If your tooling is from the same time none of them are relevant anymore if you'd ask me.
I can't help with a stack that old, sorry, you'll need to figure this out on your own.
I would recommend using rollup instead of terser and babel...

**Reply by andreapernici** [reply]
Gulp and Terser are at the latest version.

So the stack isn't old.

The code works with the latest version without minification. When minified it breaks. So I'd only like to know if there are some known specs for terser and gulp in order to work with maplibre-gl latest.

**Comment by richard-ling** [comment]
Late to this party, but have just run into this issue using maplibre-gl 4.7.1.
It was caused by the use of `Function.prototype.toString()` in the bundle prelude, together with code minification.
Terser combined `dist/maplibre-gl.js` with other modules, then created the identifier `s` for global `fetch` so in `dist/maplibre-gl.js`, `if(fetch&&Request&&AbortController&&...` minifies to `if(s&&Request&&AbortController&&...`.

The minified code can access `s` from outer scope, but the bundle prelude gets the minified source code as a string, creates a blob, and the blob is loaded into a worker in a new context with no outer scope. So, there's a crash on the worker thread ('s is not defined').

I'm not expert enough in any of this to work out a solution yet. But I guess one is to change the minifier setting to not minify `maplibre-gl` again, since its already minified.

Might be worth to put a note in the docs how to safely use code minifiers with maplibre.
